test/infrastructure/docker/controllers/dockermachine_controller.go

```diff
@@ -209,34 +232,48 @@ func (r *DockerMachineReconciler) reconcileNormal(ctx context.Context, machine *
 		dockerMachine.Status.LoadBalancerConfigured = true
 	}
 
+	// At, this stage, we are ready for bootstrap. However, if the BootstrapExecSucceededCondition is missing we add it and we
+	// requeue so the user can se the change of state before the bootstrap actually start.
+	if !conditions.Has(dockerMachine, infrav1.BootstrapExecSucceededCondition) {
+		conditions.MarkFalse(dockerMachine, infrav1.BootstrapExecSucceededCondition, infrav1.BootstrappingReason, clusterv1.ConditionSeverityInfo, "")
+		return ctrl.Result{Requeue: true}, nil
+	}
+
 	// if the machine isn't bootstrapped, only then run bootstrap scripts
 	if !dockerMachine.Spec.Bootstrapped {
 		bootstrapData, err := r.getBootstrapData(ctx, machine)
 		if err != nil {
 			r.Log.Error(err, "failed to get bootstrap data")
-			return ctrl.Result{}, nil
+			return ctrl.Result{}, err
 		}
 
 		timeoutctx, cancel := context.WithTimeout(ctx, 3*time.Minute)
 		defer cancel()
 		// Run the bootstrap script. Simulates cloud-init.
 		if err := externalMachine.ExecBootstrap(timeoutctx, bootstrapData); err != nil {
-			return ctrl.Result{}, errors.Wrap(err, "failed to exec DockerMachine bootstrap")
+			conditions.MarkFalse(dockerMachine, infrav1.BootstrapExecSucceededCondition, infrav1.BootstrapFailedReason, clusterv1.ConditionSeverityWarning, "Repeating bootstrap")
+			return ctrl.Result{RequeueAfter: 5 * time.Second}, errors.Wrap(err, "failed to exec DockerMachine bootstrap")
 		}
 		dockerMachine.Spec.Bootstrapped = true
 	}
 
+	// Update the BootstrapExecSucceededCondition condition
+	if dockerMachine.Spec.Bootstrapped {
```

Comment:
Same as above (using a separated code branch to get the code re-entrant vs setting on ExecBootstrap succeded only)

Reply:
I don't think we should be using the conditions as a mechanism for re-entrancy. They exist in status, so should be re-create-able.

Reply:
Sorry if my explanation created confusion, but we are not using conditions for re-entrancy here.
However, we are ensuring the condition get always set and not only during the reconciliation that executes the bootstrap code (condition is re-create-able)

Reply:
Same as my other comment though, isn't this already unreachable if `dockerMachine.Spec.Bootstrapped` isn't true?